src/db/mysql/batch.rs

```diff
@@ -0,0 +1,272 @@
+            serde_json::to_string(bso).map_err(|e| {
+                DbError::internal(&format!("Couldn't serialize batch::create bso: {}", e))
+            })
+        }).collect();
+    batch_strings.map(|bs| {
+        format!(
+            "{}{}",
+            bs.join("\n"),
+            if bsos.is_empty() { "" } else { "\n" }
+        )
+    })
+}
+
+#[macro_export]
+macro_rules! batch_db_method {
+    ($name:ident, $batch_name:ident, $type:ident) => {
+        pub fn $name(&self, params: params::$type) -> Result<results::$type> {
+            batch::$batch_name(self, params)
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::{
+        super::test::{db, gbso, hid, postbso},
+        *,
+    };
+    use db::params;
+
+    fn cb(user_id: u32, coll: &str, bsos: Vec<params::PostCollectionBso>) -> params::CreateBatch {
+        params::CreateBatch {
+            user_id: hid(user_id),
+            collection: coll.to_owned(),
+            bsos,
+        }
+    }
+
+    fn vb(user_id: u32, coll: &str, id: i64) -> params::ValidateBatch {
+        params::ValidateBatch {
+            user_id: hid(user_id),
+            collection: coll.to_owned(),
+            id,
+        }
+    }
+
+    fn ab(
+        user_id: u32,
+        coll: &str,
+        id: i64,
+        bsos: Vec<params::PostCollectionBso>,
+    ) -> params::AppendToBatch {
+        params::AppendToBatch {
+            user_id: hid(user_id),
+            collection: coll.to_owned(),
+            id,
+            bsos,
+        }
+    }
+
+    fn gb(user_id: u32, coll: &str, id: i64) -> params::GetBatch {
+        params::GetBatch {
+            user_id: hid(user_id),
+            collection: coll.to_owned(),
+            id,
+        }
+    }
+
+    #[test]
+    fn create_delete() -> Result<()> {
+        let db = db()?;
+
+        let uid = 1;
+        let coll = "clients";
+        let id = create(&db, cb(uid, coll, vec![]))?;
+        assert!(validate(&db, vb(uid, coll, id))?);
+        assert!(!validate(&db, vb(uid, coll, id + 1000))?);
+
+        delete(
+            &db,
+            params::DeleteBatch {
+                user_id: hid(uid),
+                collection: coll.to_owned(),
+                id,
+            },
+        )?;
+        assert!(!validate(&db, vb(uid, coll, id))?);
+        Ok(())
+    }
+
+    #[test]
+    fn update() -> Result<()> {
+        let db = db()?;
+
+        let uid = 1;
+        let coll = "clients";
+        let id = create(&db, cb(uid, coll, vec![]))?;
+        let batch = get(&db, gb(uid, coll, id))?.unwrap();
+        assert_eq!(batch.bsos, "".to_owned());
+
+        let bsos = vec![
+            postbso("b0", Some("payload 0"), Some(10), None),
+            postbso("b1", Some("payload 1"), Some(1000000000), None),
+        ];
+        append(&db, ab(uid, coll, id, bsos))?;
+
+        let batch = get(&db, gb(uid, coll, id))?.unwrap();
+        assert_ne!(batch.bsos, "".to_owned());
+        // XXX: expiry updated?
```

Comment:
I don't have strong opinions about this, but my first instance is that expiry should not be updated.  Otherwise there's no finite bound on the ultimate lifetime of a batch, a misbehaving client could slow-drip add items to a batch and keep it alive indefinitely.